### PR TITLE
[Kernel] Selectively dequantize FP8 KV for BF16 DSA prefill

### LINKED ISSUE #36338: [Feature] Selective FP8 KV dequantization for BF16 DSA sparse prefill
### Checklist

- [x] If this is not a feature request but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Motivation

In [dsa_backend.py](https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/layers/attention/dsa_backend.py), when the KV cache is stored in FP8 but `flashmla_sparse` is used to compute sparse attention in BF16, the prefix path currently dequantizes the entire flattened KV sequence:

```python
# bf16 path (dsa_impl == "flashmla_sparse").
if _has_prefix:
    page_table_1_flattened = (
        self.forward_metadata.page_table_1_flattened
    )
    assert page_table_1_flattened is not None
    kv_cache = dequantize_k_cache_paged(
        kv_cache, page_table_1_flattened
    )
else:
    kv_cache = _cat([k, k_rope], dim=-1)
```

However, DSA only attends to the top-k KV tokens selected for each query token. Therefore, especially for long contexts, dequantizing the whole sequence may introduce unnecessary dequantization work, memory traffic, and BF16 staging memory.

There is another consideration for sparse prefill. A prefill chunk contains multiple query tokens, and their top-k selections may have many overlapping KV positions. A naive per-query selective dequantization implementation could therefore dequantize the same KV position multiple times.

### Feature request

Could we add an optimized BF16 path that only dequantizes the required KV positions?

For example, it could construct the union of the top-k positions selected by all query tokens in the prefill chunk, dequantize each unique KV position once into a compact BF16 buffer, and remap the top-k indices to this compact buffer.

Since deduplication and index remapping also introduce overhead, a heuristic or fallback to the current full-sequence dequantization path may be needed when selective dequantization is not beneficial.

The goal is to keep the memory benefit of FP8 KV-cache storage while still performing sparse attention computation in BF16, without paying the cost of dequantizing the whole sequence.

This is different from the native FP8 `flashmla_sparse_q8` path, because the requested path still performs the attention computation in BF16.

Would this be feasible while maintaining good end-to-end sparse prefill performance?

### Related resources

- \#11655 introduced the current FP8-to-BF16 paged dequantization path.
- \#30514 added native FP8 sparse prefill, which avoids the BF16 staging path but uses FP8 attention computation.

### PR DESCRIPTION
**Pull request by ywgrit**
## Motivation

When FP8 DSA KV cache is consumed by the BF16 `flashmla_sparse` prefill path, SGLang currently dequantizes the complete flattened prefix. DSA only attends to selected top-k positions, and selections from a prefill chunk can overlap, so full-prefix BF16 staging wastes memory traffic and workspace for long contexts.

Closes #36338.

## What this draft implements

- Maps logical DSA top-k positions through the page table to physical KV slots.
- Deduplicates physical slots with a persistent epoch table and deterministic device-side compaction.
- Remaps top-k indices to a compact BF16 buffer.
- Extends paged FP8 KV dequantization with reusable output and a device-side valid-row extent, avoiding a host synchronization on `num_unique`.
- Bounds the CUDA Graph-safe dequant launch by SM count and consumes device-sized active rows with a grid-stride loop instead of launching one masked program per workspace-capacity row.
- Reuses grow-only metadata/BF16 buffers and covers CUDA Graph capture/replay.
- Preserves the existing full-prefix path by default. Both experimental modes are opt-in.
- Adds a conservative H20-calibrated crossover gate so short-prefix regressions fall back to full dequantization.
- Adds reproducible JSON/CSV shape benchmarks.

## H20 results

GPU: 1x NVIDIA H20, SM90, 97,871 MiB, 78 SMs  
Dtype: FP8 E4M3 KV input, BF16 compact output  
Metric: uninstrumented CUDA-event operator-path latency

Final long-prefix measurements use 50 warmups and 500 measured iterations:

| Prefix rows | Query tokens | Top-k | Physical unique rows | Cached full dequant | Dense selective total | Speedup |
|---:|---:|---:|---:|---:|---:|---:|
| 65,536 | 16 | 2,048 | 4,096 | 224.59 us | 104.41 us | 2.151x |
| 65,536 | 16 | 2,048 | 2,048 | 224.66 us | 99.23 us | 2.264x |

For the 4,096-unique-row case, replacing the capacity-sized masked launch with the SM-bounded device-extent kernel reduced isolated selected-row dequantization from 100.94 us to 28.33 us (3.56x).

Earlier crossover measurements showed that the complete selective pipeline is not profitable for short prefixes: approximately 0.28x/0.53x/0.80x/1.07x at 8K/16K/24K/32K prefix rows and 1.31x/1.58x/1.90x at 40K/48K/56K. The current experimental policy therefore fails closed below 40K and retains the conservative byte-traffic/pool-scan check above it.

CUDA allocator peak increment for the 65,536-row / 16-query case decreased from 75,497,472 bytes for full BF16 staging to 38,799,360 bytes for selective workspace plus metadata, a 48.6% reduction. The fixed CUDA-Graph-safe BF16 output capacity decreased from 65,536 to 32,768 rows; the measured physical union contained 4,096 rows.

Nsight Systems 2025.1.1 identified the original capacity-masked paged FP8 dequantization as 79.1% of captured GPU kernel time. After the follow-up, `_dequantize_k_cache_paged_device_extent_kernel` averaged 8.586 us under tracing. Nsight tracing perturbs this multi-launch microbenchmark, so speedups above are from uninstrumented CUDA-event measurements. Nsight Compute hardware counters were unavailable because the rental host disables performance-counter access for both container users and container root.

The repository's separate native Q8KV8 benchmark was also run on this H20. Its FP8 attention-only path was 1.57-1.69x faster than Q16 BF16 FlashMLA over the five official shapes. That is not directly comparable to the staging-only measurements above: this draft specifically optimizes configurations that retain BF16 sparse attention while storing KV in FP8; native Q8KV8 remains an explicit SM90-only prefill backend.

These are operator/backend-path results, not end-to-end model claims.

## Validation

- `67 passed` on H20 across the five focused kernel/unit files, covering metadata, physical alias dedup/remap, exact FP8-to-BF16 reconstruction, wrapper contracts, and combined CUDA Graph replay.
- `1 passed` on H20 through the real DSA backend: FP8 KV cache -> ragged top-k -> selective physical compaction/dequantization -> BF16 FlashMLA sparse prefill -> independent attention reference.
- `2 passed` for native Q8KV8 real-kernel parity and repeated-launch stability, used only as a scope/baseline check.
- Black 26.1.0, isort 7.0.0, Ruff 0.15.1, registered-test validation, `py_compile`, wrapper contract validation, and `git diff --check` pass.
- The latest upstream lint workflow passed before the device-extent follow-up; the amended commit will trigger it again.

## Draft limitations / promotion gates

- No end-to-end DeepSeek DSA TTFT, prefill latency, throughput, or real-workload top-k overlap distribution yet.
- The available full-model fixture uses TP8 and reports about 80.4 GiB peak-before-load per GPU, so one H20 cannot supply the full-model gate. A reduced-layer TP integration can run on fewer GPUs, while an original-model result needs 8x96GB-class GPUs.
- Nsight Compute DRAM/occupancy counters require a host that permits GPU performance-counter access.
- The 40K floor and 16x-SM launch bound are deliberately conservative H20-calibrated heuristics. They should be replaced or generalized with a cross-device/shape cost model before default enablement.
- The feature remains default-off until real-model evidence and maintainer feedback justify promotion.

---
### CI States

Latest PR Test (Base): :x: [Run #32938324700](https://github.com/sgl-project/sglang/actions/runs/32938324700)
Latest PR Test (Extra): :x: [Run #32938324446](https://github.com/sgl-project/sglang/actions/runs/32938324446)
Latest PR Test (AMD ROCm 7.2): :x: [Run #32938324711](https://github.com/sgl-project/sglang/actions/runs/32938324711)

### CODE REVIEW

**Comment by linkedlist771** [comment]
Do you implement this PR with this comment all using the Codex? It seems hard to review. I don't think it would need that much code, here is a minimal code I am trying to use:
```python


def _remap_ragged_topk_to_unique_kv_slots(
    page_table_1_flattened: torch.Tensor,
    topk_indices: torch.Tensor,
    kv_token_capacity: int,
) -> Tuple[torch.Tensor, torch.Tensor]:
    """Map ragged logical top-k indices to deduplicated physical KV slots."""
    flat_topk_indices = topk_indices.reshape(-1)
    local_page_table = torch.full_like(flat_topk_indices, -1)
    if page_table_1_flattened.numel() == 0 or flat_topk_indices.numel() == 0:
        return (
            page_table_1_flattened.new_empty((0,)),
            local_page_table.view_as(topk_indices),
        )

    logical_capacity = page_table_1_flattened.numel()
    valid_logical_mask = (flat_topk_indices >= 0) & (
        flat_topk_indices < logical_capacity
    )
    safe_logical_indices = torch.where(
        valid_logical_mask,
        flat_topk_indices,
        torch.zeros_like(flat_topk_indices),
    )
    selected_physical_slots = page_table_1_flattened[safe_logical_indices]
    valid_mask = (
        valid_logical_mask
        & (selected_physical_slots >= 0)
        & (selected_physical_slots < kv_token_capacity)
    )

    unique_physical_slots, inverse_indices = torch.unique(
        selected_physical_slots[valid_mask], sorted=True, return_inverse=True
    )
    local_page_table[valid_mask] = inverse_indices.to(local_page_table.dtype)
    return unique_physical_slots.contiguous(), local_page_table.view_as(topk_indices)


def _dequantize_selected_ragged_kv_cache(
    kv_cache: torch.Tensor,
    page_table_1_flattened: torch.Tensor,
    topk_indices: torch.Tensor,
) -> Tuple[torch.Tensor, torch.Tensor]:
    """Dequantize each physical KV slot selected by ragged top-k exactly once.

    This stays on eager prefill because ``torch.unique`` has a data-dependent shape.
    FlashMLA still needs one addressable KV row when every top-k entry is invalid.
    """
    kv_token_capacity = kv_cache.numel() // kv_cache.shape[-1]
    unique_physical_slots, local_page_table = (
        _remap_ragged_topk_to_unique_kv_slots(
            page_table_1_flattened, topk_indices, kv_token_capacity
        )
    )
    selected_kv_cache = dequantize_k_cache_paged(kv_cache, unique_physical_slots)
    if selected_kv_cache.shape[0] == 0:
        selected_kv_cache = selected_kv_cache.new_zeros(
            (1, *selected_kv_cache.shape[1:])
        )
    return selected_kv_cache, local_page_table

```

**Comment by linkedlist771** [comment]
I don't think it is a good idea to copy paste my issue description to Codex with:
```bash

/goal hi, Codex, this is the issue description, please implement it and draft a PR.
````

**Comment by ywgrit** [comment]
> Do you implement this PR with this comment all using the Codex? It seems hard to review. I don't think it would need that much code, here is a minimal code I am trying to use:
> 
> ```python
> def _remap_ragged_topk_to_unique_kv_slots(
>     page_table_1_flattened: torch.Tensor,
>     topk_indices: torch.Tensor,
>     kv_token_capacity: int,
> ) -> Tuple[torch.Tensor, torch.Tensor]:
>     """Map ragged logical top-k indices to deduplicated physical KV slots."""
>     flat_topk_indices = topk_indices.reshape(-1)
>     local_page_table = torch.full_like(flat_topk_indices, -1)
>     if page_table_1_flattened.numel() == 0 or flat_topk_indices.numel() == 0:
>         return (
>             page_table_1_flattened.new_empty((0,)),
>             local_page_table.view_as(topk_indices),
>         )
> 
>     logical_capacity = page_table_1_flattened.numel()
>     valid_logical_mask = (flat_topk_indices >= 0) & (
>         flat_topk_indices < logical_capacity
>     )
>     safe_logical_indices = torch.where(
>         valid_logical_mask,
>         flat_topk_indices,
>         torch.zeros_like(flat_topk_indices),
>     )
>     selected_physical_slots = page_table_1_flattened[safe_logical_indices]
>     valid_mask = (
>         valid_logical_mask
>         & (selected_physical_slots >= 0)
>         & (selected_physical_slots < kv_token_capacity)
>     )
> 
>     unique_physical_slots, inverse_indices = torch.unique(
>         selected_physical_slots[valid_mask], sorted=True, return_inverse=True
>     )
>     local_page_table[valid_mask] = inverse_indices.to(local_page_table.dtype)
>     return unique_physical_slots.contiguous(), local_page_table.view_as(topk_indices)
> 
> 
> def _dequantize_selected_ragged_kv_cache(
>     kv_cache: torch.Tensor,
>     page_table_1_flattened: torch.Tensor,
>     topk_indices: torch.Tensor,
> ) -> Tuple[torch.Tensor, torch.Tensor]:
>     """Dequantize each physical KV slot selected by ragged top-k exactly once.
> 
>     This stays on eager prefill because ``torch.unique`` has a data-dependent shape.
>     FlashMLA still needs one addressable KV row when every top-k entry is invalid.
>     """
>     kv_token_capacity = kv_cache.numel() // kv_cache.shape[-1]
>     unique_physical_slots, local_page_table = (
>         _remap_ragged_topk_to_unique_kv_slots(
>             page_table_1_flattened, topk_indices, kv_token_capacity
>         )
>     )
>     selected_kv_cache = dequantize_k_cache_paged(kv_cache, unique_physical_slots)
>     if selected_kv_cache.shape[0] == 0:
>         selected_kv_cache = selected_kv_cache.new_zeros(
>             (1, *selected_kv_cache.shape[1:])
>         )
>     return selected_kv_cache, local_page_table
> ```

  Yes, I used Codex heavily to implement and test this draft. The decision to include the custom compaction kernel, static workspace, CUDA Graph support, heuristics, and benchmark infrastructure in the first version made the PR unnecessarily large and difficult to review. I should have started with the minimal eager implementation.
  Your `torch.unique` approach is much smaller and directly addresses the core functional gap. If you think a PR scoped to that minimal implementation would be acceptable and worth reviewing, I am willing to refactor this draft accordingly: remove the dense-epoch compaction kernel, static workspace/CUDA-Graph-specific machinery, and extensive benchmark infrastructure, while retaining only the selected-KV dequantization path and focused correctness/backend tests.
  If you are already planning to submit this implementation yourself, please let me know and I will avoid duplicating the work. I will wait for your direction before rewriting the PR.

**Comment by linkedlist771** [comment]
> Yes, I used Codex heavily to implement and test this draft. The decision to include the custom compaction kernel, static workspace, CUDA Graph support, heuristics, and benchmark infrastructure in the first version made the PR unnecessarily large and difficult to review. I should have started with the minimal eager implementation.
> 
> Your `torch.unique` approach is much smaller and directly addresses the core functional gap. If you think a PR scoped to that minimal implementation would be acceptable and worth reviewing, I am willing to refactor this draft accordingly: remove the dense-epoch compaction kernel, static workspace/CUDA-Graph-specific machinery, and extensive benchmark infrastructure, while retaining only the selected-KV dequantization path and focused correctness/backend tests.
> 
> If you are already planning to submit this implementation yourself, or if the team does not want this change, please let me know and I will avoid duplicating the work. I will wait for your direction before rewriting the PR.

We may wait for the maintainer's decision about this issue.
